src/browser-installer/registry/index.ts

```diff
@@ -0,0 +1,215 @@
+
+export const getBinaryPath = async (name: BinaryName, platform: BrowserPlatform, version: string): Promise<string> => {
+    const registryKey = getRegistryKey(name, platform);
+
+    if (!registry[registryKey]) {
+        throw new Error(`Binary '${name}' on '${platform}' is not installed`);
+    }
+
+    if (!registry[registryKey][version]) {
+        throw new Error(`Version '${version}' of driver '${name}' on '${platform}' is not installed`);
+    }
+
+    const binaryRelativePath = await registry[registryKey][version];
+
+    browserInstallerDebug(`resolved '${name}@${version}' on ${platform} to ${binaryRelativePath}`);
+
+    return path.resolve(registryPath, binaryRelativePath);
+};
+
+const addBinaryToRegistry = (
+    name: BinaryName,
+    platform: BrowserPlatform,
+    version: string,
+    absoluteBinaryPath: string,
+): void => {
+    const registryKey = getRegistryKey(name, platform);
+    const relativePath = path.relative(registryPath, absoluteBinaryPath);
+
+    registry[registryKey] ||= {};
+    registry[registryKey][version] = relativePath;
+
+    const replacer = (_: string, value: unknown): unknown | undefined => {
+        if ((value as Promise<unknown>).then) {
+            return;
+        }
+
+        return value;
+    };
+
+    browserInstallerDebug(`adding '${name}@${version}' on '${platform}' to registry at ${relativePath}`);
+    outputJSONSync(registryPath, registry, { replacer });
+};
+
+const getBinaryVersions = (name: BinaryName, platform: BrowserPlatform): string[] => {
+    const registryKey = getRegistryKey(name, platform);
+
+    if (!registry[registryKey]) {
+        return [];
+    }
+
+    return Object.keys(registry[registryKey]);
+};
+
+const hasBinaryVersion = (name: BinaryName, platform: BrowserPlatform, version: string): boolean =>
+    getBinaryVersions(name, platform).includes(version);
+
+export const getMatchingDriverVersion = (
+    driverName: SupportedDriver,
+    platform: BrowserPlatform,
+    browserVersion: string,
+): string | null => {
+    const registryKey = getRegistryKey(driverName, platform);
+
+    if (!registry[registryKey]) {
+        return null;
+    }
+
+    if (driverName === Driver.CHROMEDRIVER || driverName === Driver.EDGEDRIVER) {
+        const milestone = getMilestone(browserVersion);
+        const buildIds = getBinaryVersions(driverName, platform);
+        const suitableBuildIds = buildIds.filter(buildId => buildId.startsWith(milestone));
+
+        if (!suitableBuildIds.length) {
+            return null;
+        }
+
+        const suitableBuildIdsSorted = suitableBuildIds.sort(semverVersionsComparator);
+
+        return suitableBuildIdsSorted[suitableBuildIdsSorted.length - 1];
+    }
+
+    if (driverName === Driver.GECKODRIVER) {
+        const buildIds = Object.keys(registry[registryKey]);
+        const buildIdsSorted = buildIds.sort(semverVersionsComparator);
+
+        return buildIdsSorted.length ? buildIdsSorted[buildIdsSorted.length - 1] : null;
+    }
```

Comment:
geckodriver is compatible with all browser versions, so we just pick the most recent version for firefox